File: ios/App/App/IsolatedModules/FileExplorer.swift
//
//  FileExplorer.swift
//  App
//
//  Created by Julia Samol on 08.02.23.
//

import Foundation

// MARK: FileExplorer

struct FileExplorer {
    static let shared: FileExplorer = .init()
    
    private let assets: AssetsExplorer
    private let document: DocumentExplorer
    
    init(fileManager: FileManager = .default) {
        self.assets = .init(fileManager: fileManager)
        self.document = .init(fileManager: fileManager)
    }
    
    func readIsolatedModulesScript() throws -> String {
        try assets.readIsolatedModulesScript()
    }
    
    func loadAssetModules() throws -> [JSModule] {
        try loadModules(explorer: assets).map { .asset($0) }
    }
    
    func loadExternalModules() throws -> [JSModule] {
        try loadModules(explorer: document).map { .external($0) }
    }
    
    func readModuleSources(_ module: JSModule) throws -> [String] {
        switch module {
        case .asset(let asset):
            return try assets.readModuleSources(asset)
        case .external(let external):
            return try document.readModuleSources(external)
        }
    }
    
    private func loadModules<T: JSModuleProtocol, E: DynamicSourcesExplorer>(explorer: E) throws -> [T] where E.T == T {
        try explorer.listModules().map { module in
            let jsonDecoder = JSONDecoder()
            guard let manifestData = try explorer.readModuleManifest(module).data(using: .utf8) else {
                throw JSError.invalidJSON
            }
            
            let manifest = try jsonDecoder.decode(ModuleManifest.self, from: manifestData)
            let namespace = manifest.src?.namespace
            let preferredEnvironment = manifest.jsenv?.ios ?? .webview
            let paths = try manifest.include.map { try explorer.absoluteModulePath(ofPath: "\(module)/\($0)") }
            
            return T.init(identifier: module, namespace: namespace, preferredEnvironment: preferredEnvironment, paths: paths)
        }
    }
}

// MARK: AssetsExplorer

private struct AssetsExplorer: DynamicSourcesExplorer {
    typealias T = JSModule.Asset
    
    static let assetsURL: URL = Bundle.main.url(forResource: "public", withExtension: nil)!.appendingPathComponent("assets")
    private static let script: String = "native/isolated_modules/isolated-modules.script.js"
    private static let modulesDir: String = "protocol_modules"
    
    private let fileManager: FileManager
    
    init(fileManager: FileManager) {
        self.fileManager = fileManager
    }
    
    func readIsolatedModulesScript() throws -> String {
        try readString(atPath: Self.script)
    }
    
    func listModules() throws -> [String] {
        let url = Self.assetsURL.appendingPathComponent(Self.modulesDir)
        return try fileManager.contentsOfDirectory(atPath: url.path)
    }
    
    func absoluteModulePath(ofPath path: String) throws -> String {
        "\(Self.modulesDir)/\(path)"
    }
    
    func readModuleSources(_ module: JSModule.Asset) throws -> [String] {
        try module.paths.lazy.map { try readString(atPath: $0) }
    }
    
    func readModuleManifest(_ module: String) throws -> String {
        try readString(atPath: "\(absoluteModulePath(ofPath: module))/manifest.json")
    }
    
    private func readString(atPath pathComponent: String) throws -> String {
        let url = Self.assetsURL.appendingPathComponent(pathComponent)
        return try fileManager.stringContents(atPath: url.path)
    }
}

// MARK: DocumentExplorer

private struct DocumentExplorer: DynamicSourcesExplorer {
    typealias T = JSModule.External
    
    private static let modulesDir: String = "protocol_modules"
    
    private let fileManager: FileManager
    
    init(fileManager: FileManager) {
        self.fileManager = fileManager
    }
    
    func listModules() throws -> [String] {
        guard let url = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first else {
            return []
        }
        
        let modulesDirPath = url.appendingPathComponent(Self.modulesDir).path
        guard fileManager.fileExists(atPath: modulesDirPath) else {
            return []
        }
        
        return try fileManager.contentsOfDirectory(atPath: modulesDirPath)
    }
    
    func absoluteModulePath(ofPath path: String) throws -> String {
        return "\(Self.modulesDir)/\(path)"
    }
    
    func readModuleSources(_ module: JSModule.External) throws -> [String] {
        try module.paths.lazy.map { try readString(atPath: $0) }
    }
    
    func readModuleManifest(_ module: String) throws -> String {
        try readString(atPath: "\(absoluteModulePath(ofPath: module))/manifest.json")
    }
    
    private func readString(atPath path: String) throws -> String {
        let url = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent(path)
        return try fileManager.stringContents(atPath: url.path)
    }
}

// MARK: DynamicSourcesExplorer

private protocol DynamicSourcesExplorer {
    associatedtype T
    
    func listModules() throws -> [String]
    func absoluteModulePath(ofPath path: String) throws -> String
    
    func readModuleSources(_ module: T) throws -> [String]
    func readModuleManifest(_ module: String) throws -> String
}

// MARK: Extensions

private extension FileManager {
    func stringContents(atPath path: String) throws -> String {
        guard let data = contents(atPath: path),
              let string = String(data: data, encoding: .utf8) else {
            throw Error.invalidPath
        }
        
        return string
    }
}

private enum Error: Swift.Error {
    case invalidPath
}

File: android/app/src/main/java/it/airgap/vault/plugin/isolatedmodules/FileExplorer.kt
package it.airgap.vault.plugin.isolatedmodules

import android.content.Context
import com.getcapacitor.JSObject
import it.airgap.vault.plugin.isolatedmodules.js.JSModule
import it.airgap.vault.plugin.isolatedmodules.js.environment.JSEnvironment
import it.airgap.vault.util.readBytes
import java.io.File

interface StaticSourcesExplorer {
    fun readJavaScriptEngineUtils(): ByteArray
    fun readIsolatedModulesScript(): ByteArray
}

interface DynamicSourcesExplorer<in M : JSModule> {
    fun listModules(): List<String>
    fun absoluteModulePath(path: String): String

    fun readModuleSources(module: M): Sequence<ByteArray>
    fun readModuleManifest(module: String): ByteArray
}

class FileExplorer private constructor(
    private val assetsExplorer: AssetsExplorer,
    private val internalFilesExplorer: InternalFilesExplorer,
) : StaticSourcesExplorer by assetsExplorer {
    constructor(context: Context) : this(AssetsExplorer(context), InternalFilesExplorer(context))

    fun loadAssetModules(): List<JSModule> = loadModules(assetsExplorer, JSModule::Asset)
    fun loadExternalModules(): List<JSModule> = loadModules(internalFilesExplorer, JSModule::External)

    fun readModuleSources(module: JSModule): Sequence<ByteArray> =
        when (module) {
            is JSModule.Asset -> assetsExplorer.readModuleSources(module)
            is JSModule.External -> internalFilesExplorer.readModuleSources(module)
        }

    private fun <T : JSModule> loadModules(
        explorer: DynamicSourcesExplorer<T>,
        constructor: (identifier: String, namespace: String?, preferredEnvironment: JSEnvironment.Type, paths: List<String>) -> T,
    ): List<T> = explorer.listModules().map { module ->
        val manifest = JSObject(explorer.readModuleManifest(module).decodeToString())
        val namespace = manifest.getJSObject("src")?.getString("namespace")
        val preferredEnvironment = manifest.getJSObject("jsenv")?.getString("android")?.let { JSEnvironment.Type.fromString(it) } ?: JSEnvironment.Type.JavaScriptEngine
        val paths = buildList {
            val include = manifest.getJSONArray("include")
            for (i in 0 until include.length()) {
                val path = include.getString(i).takeIf { it.endsWith(".js") } ?: continue
                add(explorer.absoluteModulePath("$module/${path.trimStart('/')}"))
            }
        }

        constructor(module, namespace, preferredEnvironment, paths)
    }
}

private class AssetsExplorer(private val context: Context) : StaticSourcesExplorer, DynamicSourcesExplorer<JSModule.Asset> {
    override fun readJavaScriptEngineUtils(): ByteArray = context.assets.readBytes(JAVA_SCRIPT_ENGINE_UTILS)
    override fun readIsolatedModulesScript(): ByteArray = context.assets.readBytes(SCRIPT)

    override fun listModules(): List<String> = context.assets.list(MODULES_DIR)?.toList() ?: emptyList()
    override fun absoluteModulePath(path: String): String = "${MODULES_DIR}/${path.trimStart('/')}"

    override fun readModuleSources(module: JSModule.Asset): Sequence<ByteArray> = module.paths.asSequence().map { context.assets.readBytes(it)}
    override fun readModuleManifest(module: String): ByteArray = context.assets.readBytes("${absoluteModulePath(module)}/manifest.json")

    companion object {
        private const val SCRIPT = "public/assets/native/isolated_modules/isolated-modules.script.js"
        private const val JAVA_SCRIPT_ENGINE_UTILS = "public/assets/native/isolated_modules/isolated-modules.js-engine-android.js"

        private const val MODULES_DIR = "public/assets/protocol_modules"
    }
}

private class InternalFilesExplorer(private val context: Context) : DynamicSourcesExplorer<JSModule.External> {
    private val modulesDir: File
        get() = File(context.filesDir, MODULES_DIR)

    override fun listModules(): List<String> = modulesDir.list()?.toList() ?: emptyList()
    override fun absoluteModulePath(path: String): String = File(modulesDir, path).absolutePath

    override fun readModuleSources(module: JSModule.External): Sequence<ByteArray> = module.paths.asSequence().map { File(it).readBytes() }
    override fun readModuleManifest(module: String): ByteArray = File(modulesDir, "$module/manifest.json").readBytes()

    companion object {
        private const val MODULES_DIR = "protocol_modules"
    }
}